src/IO/S3Common.cpp

```diff
@@ -91,28 +107,103 @@ class AWSLogger final : public Aws::Utils::Logging::LogSystemInterface
     std::unordered_map<String, Poco::Logger *> tag_loggers;
 };
 
+class AWSSTSAssumeRoleCredentialsProvider : public Aws::Auth::AWSCredentialsProvider
+{
+    static const int ACCOUNT_FOR_LATENCY = 60;
+
+public:
+    AWSSTSAssumeRoleCredentialsProvider(std::shared_ptr<Aws::STS::STSClient> sts_client_):
+        sts_client(std::move(sts_client_)),
+        session_name("aws-sdk-cpp-" + std::to_string(Aws::Utils::DateTime::CurrentTimeMillis()))
+    {
+    }
+
+    Aws::Auth::AWSCredentials GetAWSCredentials() override
+    {
+        auto diff_seconds = static_cast<int>(Aws::Utils::DateTime::Now().SecondsWithMSPrecision() - Aws::Utils::DateTime(expiry.load()).SecondsWithMSPrecision());
+        if (diff_seconds > 0 - ACCOUNT_FOR_LATENCY)
+        {
+            auto * logger = &Poco::Logger::get("STSAssumeRoleCredentialsProvider");
+            LOG_DEBUG(logger, "Credentials have expired with diff of {} since last credentials pull.", diff_seconds);
+            LOG_TRACE(logger, "Grabbing lock.");
+            std::lock_guard<std::mutex> locker(credentials_mutex);
+            LOG_TRACE(logger, "Lock acquired. Checking expiration again.");
+            diff_seconds = static_cast<int>(Aws::Utils::DateTime::Now().SecondsWithMSPrecision() - Aws::Utils::DateTime(expiry.load()).SecondsWithMSPrecision());
+            if (diff_seconds > 0 - ACCOUNT_FOR_LATENCY)
+            {
+                auto caller_identity_outcome = sts_client->GetCallerIdentity({});
+                if (caller_identity_outcome.IsSuccess())
+                {
+                    String role_arn = caller_identity_outcome.GetResult().GetArn();
+                    if (role_arn != cached_role_arn || !underlying_credentials_provider)
+                    {
+                        underlying_credentials_provider = std::make_shared<Aws::Auth::STSAssumeRoleCredentialsProvider>(
+                            role_arn, session_name, String(), Aws::Auth::DEFAULT_CREDS_LOAD_FREQ_SECONDS, sts_client
+                        );
+                        cached_role_arn = role_arn;
+                    }
+                    cached_credentials = underlying_credentials_provider->GetAWSCredentials();
+                    expiry = Aws::Utils::DateTime::Now().SecondsWithMSPrecision();
+                }
+                else
+                {
+                    LOG_ERROR(logger, "Failed to make STS GetCallerIdentity request.");
+                }
+            }
+        }
+
+        std::lock_guard<std::mutex> locker(credentials_mutex);
```

Comment:
Why is it needed here?

Reply:
It was needed to avoid race condition between read and write from `cached_credentials`